Issue #105 - Refactor `MMENSDateWrapper` into `MMEDate`

alfwatt opened the pull request:
Implement #105

Commit:
Issue #105 - Refactor `MMENSDateWrapper` into `MMEDate`

Commit:
Implement the required `NSDate` subclass overrides and protocols
Add `offsetFromServer` on `MMEDate` to track the offset at the time a date is created
Remove wrapper instances and get test running

rclee commented on MapboxMobileEvents/MMEDate.h:
```diff
@@ -0,0 +1,42 @@
+/*! @biref store the recorded time offset for the date when created using the +[MMEDate date] or +[MMEDate dateWithOffsetFromServer:(NSTimeInterval)offset]` methods */
+@property(nonatomic,assign,readonly) NSTimeInterval offsetFromServer;
+
+/*! @brief computes, reccords and returns the time offset from the server's time frame */
```
```suggestion
/*! @brief computes, records and returns the time offset from the server's time frame */
```

rclee commented on MapboxMobileEvents/MMEDate.h:
```diff
@@ -0,0 +1,42 @@
+
+@interface NSDate (MMEDate)
+
+- (NSDate *)mme_oneDayLater; // --> (instanctype) startOfTomorrow;
```
```suggestion
- (NSDate *)mme_oneDayLater; // --> (instancetype) startOfTomorrow;
```

Commit:
Tests for MMEDate, remove all DateWrapper references

alfwatt commented:
Added tests for MMEDate, deferred the system time work to #107

rclee commented on MapboxMobileEventsTests/MMEEventsManagerTests.mm:
```diff
@@ -917,9 +910,6 @@ - (void)pushEvent:(MMEEvent *)event;
-            spy_on(dateWrapper);
-            dateWrapper stub_method(@selector(formattedDateStringForDate:)).and_return(dateString);
-            eventsManager.dateWrapper = dateWrapper;
             
```
```suggestion
            NSDateFormatter *dateFormatter = MMEDate.iso8601DateFormatter;
            spy_on(dateFormatter);
            dateFormatter stub_method(@selector(stringFromDate:)).and_return(dateString);
            
```

rclee commented on MapboxMobileEventsTests/MMEEventsManagerTests.mm:
```diff
@@ -788,10 +784,7 @@ - (void)pushEvent:(MMEEvent *)event;
-            dateWrapper stub_method(@selector(formattedDateStringForDate:)).and_return(dateString);
-            eventsManager.dateWrapper = dateWrapper;
-            
+
```
(never used these suggested changes -- so, feel free to do what you want with that)

rclee reviewed (dismissed):
I think we're good to go after the tests are fixed 👍

alfwatt commented on MapboxMobileEvents/MMEDate.m:
```diff
@@ -0,0 +1,127 @@
+- (void)encodeWithCoder:(NSCoder *)aCoder {
+    [aCoder encodeDouble:_sinceReferenceDate forKey:@"sinceReferenceDate"];
+    [aCoder encodeDouble:_offsetFromServer forKey:@"offsetFromServer"];
+    [aCoder encodeInteger:1 forKey:@"MMEDateVersion"];
```
Yes, and there should be a version check

rclee committed:
Update MapboxMobileEventsTests/MMEEventsManagerTests.mm

Co-Authored-By: alfwatt <alf@istumbler.net>

rclee committed:
Update MapboxMobileEventsTests/MMEEventsManagerTests.mm

Co-Authored-By: alfwatt <alf@istumbler.net>

Commit:
Review: fix nits in .h, use constants for encode/decode, check coded version

Commit:
Consolidate Date formatters into MMEDate
MMEMetrics.date -> MMEMetrics.recordingStarted

Commit:
Don't re-declare NSSecureCoding on MMEDate

Commit:
First test for MMEDate Code/Decode

Commit:
Add more precise tests for `mme_startOfTomorrow`

Commit:
Sort Source Files by name…
Remove principal class entry from plist

Commit:
Overriding NSCoder on Foundation objects is not without peril
NSDate is a cluster and overrides `-[NSObject classForCoder]`, MMEDate now over-overrides it
Updated tests for secure coding, passing

alfwatt commented:
Tests are passing locally, ready for final review.

rclee commented on MapboxMobileEventsTests/MMEDateTests.mm:
```diff
@@ -0,0 +1,126 @@
+
+    context(@"+ recordTimeOffsetFromServer:", ^{
+        it(@"computes offsets from server time", ^{
+            NSDate* serverTime = [NSDate dateWithTimeIntervalSinceNow:interval];
```
nit: several pointer position misplacements in this file

rclee commented on MapboxMobileEventsTests/MMEMetricsManager.mm:
```diff
@@ -56,18 +56,17 @@
-                [dateFormatter dateFromString:manager.metrics.dateUTCString] should_not be_nil;
+
+            it(@"should set MMEEventDateUTC attributes with the correct format", ^{
+                NSDictionary* attributes = [manager attributes];
```
nit: pointer position

rclee commented on MapboxMobileEventsTests/MMEMetricsManager.mm:
```diff
@@ -56,18 +56,17 @@
+            it(@"should set MMEEventDateUTC attributes with the correct format", ^{
+                NSDictionary* attributes = [manager attributes];
+
+                [dateFormatter dateFromString:attributes[MMEEventDateUTC]] should_not be_nil;
```
I should have wrote this before to check the proper format "yyyy-MM-dd"...

rclee reviewed (dismissed):
Sorry, for all the nit-picks -- everything else looks solid.

Commit:
Review Nits

rclee reviewed (dismissed):
🚀

Commit:
Test improvements

alfwatt commented:
Fixed up the last few test nits as well

rclee reviewed (approved):
Thanks Alf, looks great!
